content/en/blog/_posts/2024-02-29-cri-o-seccomp-oci-artifacts.md

```diff
@@ -0,0 +1,477 @@
+the `RuntimeDefault` profile if not specified in the pod spec. For testing the
+new annotation, either disable it or lookout for the log line: _"Found container
+specific seccomp profile annotation"_.
+{{< /note >}}
+
+### Using the annotation for a container image
+
+While specifying seccomp profiles as OCI artifacts on certain workloads is a
+cool feature, the majority of end users would like to link seccomp profiles to
+actual container images. This can be done by using the same annotation, but now
+at the container image itself. For example, [Podman](https://podman.io) can be
+used to add the annotation directly on image build:
+
+```shell
+podman build \
+    --annotation seccomp-profile.kubernetes.cri-o.io/POD=quay.io/crio/seccomp:v1 \
+    -t quay.io/crio/nginx-seccomp:pod .
+```
+
+The pushed image then contains the annotation:
+
+```shell
+skopeo inspect --raw docker://quay.io/crio/nginx-seccomp:pod |
+    jq '.annotations."seccomp-profile.kubernetes.cri-o.io/POD"'
+```
+
+```txt
+"quay.io/crio/seccomp:v1"
+```
+
+If I now use that image in an CRI-O test pod definition:
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: pod
+spec:
+  containers:
+    - name: container
+      image: quay.io/crio/nginx-seccomp:pod
+```
+
+Then the CRI-O logs will indicate that the annotation got evaluated and the
+profile got applied:
+
+```shell
+kubectl delete pod/pod
+```
+
+```txt
+pod "pod" deleted
+```
+
+```shell
+kubectl apply -f pod.yaml
+```
+
+```txt
+pod/pod created
+```
+
+```txt
+INFO[…] Found image specific seccomp profile annotation for pod: seccomp-profile.kubernetes.cri-o.io=quay.io/crio/seccomp:v1  id=c1f22c59-e30e-4046-931d-a0c0fdc2c8b7 name=/runtime.v1.RuntimeService/CreateContainer
+INFO[…] Pulling OCI artifact from ref: quay.io/crio/seccomp:v1  id=c1f22c59-e30e-4046-931d-a0c0fdc2c8b7 name=/runtime.v1.RuntimeService/CreateContainer
+INFO[…] Trying to read profile from: /var/lib/containers/storage/overlay/c15f3eb1efd9abe0ae973e050c374988bba52c45a53f5c3cd80419a865013fa6/merged/seccomp.json  id=c1f22c59-e30e-4046-931d-a0c0fdc2c8b7 name=/runtime.v1.RuntimeService/CreateContainer
+INFO[…] Retrieved OCI artifact seccomp profile of len: 18853  id=c1f22c59-e30e-4046-931d-a0c0fdc2c8b7 name=/runtime.v1.RuntimeService/CreateContainer
+INFO[…] Created container 116a316cd9a11fe861dd04c43b94f45046d1ff37e2ed05a4e4194fcaab29ee63: default/pod/container  id=c1f22c59-e30e-4046-931d-a0c0fdc2c8b7 name=/runtime.v1.RuntimeService/CreateContainer
+```
+
+```shell
+export CONTAINER_ID=$(sudo crictl ps --name container -q)
+sudo crictl inspect $CONTAINER_ID | jq .info.runtimeSpec.linux.seccomp | head
+```
+
+```json
+{
+  "defaultAction": "SCMP_ACT_ERRNO",
+  "defaultErrnoRet": 38,
+  "architectures": [
+    "SCMP_ARCH_X86_64",
+    "SCMP_ARCH_X86",
+    "SCMP_ARCH_X32"
+  ],
+  "syscalls": [
+    {
+```
+
+The whole feature also works when using the container specific annotation on an
+image, for example if a container is named `container1`:
```

Comment:
- OCI image annotations aren't supposed to contain a `/` (the rules are different from Kubernetes annotations).
- How would the image author know what the container running it might be named?

Reply:
> How would the image author know what the container running it might be named?

I would not leave out that use case since it seems to be possible in small environments that the author of the pod manifest is also the image creator. But I agree this might be a rare case.

> OCI image annotations aren't supposed to contain a / (the rules are different from Kubernetes annotations).

I'm not sure if we should change the behavior for images and only allow the annotation `seccomp-profile.kubernetes.cri-o.io`. Should we? Would it be more confusing to end users?